### ISSUE TITLE
[tflite][operator: cast] How to quantize MobileNetV2 for deeplabV3+?

### ISSUE THREAD

**Issue by kanul** [tensorflow/tensorflow#20403]
**System information**
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:no
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:Ubuntu 16.04
- **TensorFlow installed from (source or binary)**:Source
- **TensorFlow version (use command below)**:1.9.0
- **Python version**:2.7.12
- **Bazel version (if compiling from source)**:0.12.0
- **GCC/Compiler version (if compiling from source)**:5.4.0
- **CUDA/cuDNN version**:cuda-9.0/7.0
- **GPU model and memory**:GeForce GTX 1080/8105MiB
- **Phone**:xiaomi5 (Snapdragon 820)
- **Exact command to reproduce**:
bazel run --config=opt //tensorflow/contrib/lite/toco:toco --
--input_file=/external_home/data/model/deeplabv3_mnv2_pascal_train_aug/frozen_inference_graph.pb
--output_file=/external_home/data/model/deeplabv3_mnv2_pascal_train_aug/kanul.tflite
--inference_type=QUANTIZED_UINT8
--input_shape=1,513,513,3
--input_array=sub_7
--output_array=logits/semantic/BiasAdd

**Describe the problem**
I have tried to quantize MobileNetV2 for deeplabV3+ with TFlite. But I fail to convert the model.
From the following issue, I saw that the operations were not supported for the option of quantization.

https://github.com/tensorflow/models/blob/master/research/deeplab/g3doc/model_zoo.md
Checkpoint name: mobilenetv2_coco_voc_trainaug

Who can explain and support to resolve the issue?

**Source code / logs**
bazel run --config=opt //tensorflow/contrib/lite/toco:toco --
--input_file=/external_home/data/model/deeplabv3_mnv2_pascal_train_aug/frozen_inference_graph.pb
--output_file=/external_home/data/model/deeplabv3_mnv2_pascal_train_aug/kanul.tflite
--inference_type=QUANTIZED_UINT8
--input_shape=1,513,513,3
--input_array=sub_7
--output_array=logits/semantic/BiasAdd
--default_ranges_min=0  
--default_ranges_max=6

Unimplemented: this graph contains an operator of type Cast for which the quantized form is not yet implemented. Sorry, and patches welcome (that's a relatively fun patch to write, mostly providing the actual quantized arithmetic code for this op).

**Source code / logs**
bazel run --config=opt //tensorflow/contrib/lite/toco:toco --
--input_file=/external_home/data/model/deeplabv3_mnv2_pascal_train_aug/frozen_inference_graph.pb
--output_file=/external_home/data/model/deeplabv3_mnv2_pascal_train_aug/kanul.tflite
--inference_type=QUANTIZED_UINT8
--input_shape=1,513,513,3
--input_array=sub_7
--output_array=logits/semantic/BiasAdd

tensorflow/contrib/lite/toco/graph_transformations/hardcode_min_max.cc:105] Tweaking the MinMax of array ResizeBilinear_1, which is an input to {Concatenation operator with output concat}, because we want all inputs and outputs of a Concatenation operator to have the same MinMax so that it can be implemented as a pure byte-copy, no arithmetic.

tensorflow/contrib/lite/toco/tflite/export.cc:367] Some of the operators in the model are not supported by the standard TensorFlow Lite runtime. If you have a custom implementation for them you can disable this error with --allow_custom_ops, or by setting allow_custom_ops=True when calling tf.contrib.lite.toco_convert(). Here is a list of operators for which  you will need custom implementations: Stack.

**Comment by karmel** [comment]
@suharshs - is there a unified listing somewhere of the quantized mobilenet models that are available with TFLite?

**Comment by bhack** [comment]
@karmel is related to https://github.com/tensorflow/tensorflow/issues/17468?

**Comment by karmel** [comment]
I suspect similar to #20272 but with mobilenet v2 instead of v1

**Comment by kanul** [comment]
@suharshs

Did you check out whether the kernel of cast is implemented or not?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 17 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by dodiknikola** [comment]
I am also wondering about the timeline for Stack.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @suharshs, @liyunlu0618: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by suharshs** [comment]
Hi all, apologies for the delay,

Stack has been implemented in this change: https://github.com/tensorflow/tensorflow/commit/df7344f1933d932f03f472402068ff1883f0c011#diff-0a7763f2b5c90990e015f610fa5f49df

**Comment by myth01** [comment]
@suharshs How about cast?

**Comment by PK15946** [comment]
Hi， @suharshs. Thanks for the implement of stack. But I have encounter a problem when use stack [here](https://github.com/tensorflow/tensorflow/issues/23855). Do you have any idea to solve it?

**Comment by PK15946** [comment]
Hi， @suharshs. Thanks for the implement of stack. But I have encounter a problem when use stack here. Do you have any idea to solve it?

**Comment by iariav** [comment]
i join @myth01 's request for the support on cast...
cannot convert a deeplabv3+ model without it..

**Comment by burui11087** [comment]
HI all,

try add --change_concat_input_ranges=false when using toco to quantize your model.
